src/main/java/io/cdap/plugin/gcp/bigquery/sink/AbstractBigQuerySink.java

```diff
@@ -129,35 +139,69 @@ protected final void initOutput(BatchSinkContext context, String outputName,
     context.addOutput(Output.of(outputName, getOutputFormatProvider(configuration, tableName, tableSchema)));
   }
 
-  protected final void decodeSimpleTypes(JsonObject json, String name, StructuredRecord input) {
-    Object object = input.get(name);
-    Schema.Field field = input.getSchema().getField(name);
-
-    if (field == null) {
-      throw new IllegalStateException(String.format("Field '%s' is absent in input record", name));
+  /**
+   * Decodes object and writes to json writer.
+   * @param writer json writer to write the object to
+   * @param name name of the field to be decoded
+   * @param object object to be decoded
+   * @param fieldSchema field schema to be decoded
+   */
+  protected void decode(JsonWriter writer, @Nullable String name, Object object,
```

Comment:
these decode methods are really about transforming a StructuredRecord into json object. It would be better if we pull out this logic from the plugins into a separate class (ex: RecordJsonTransformer). We can then have comprehensive unit tests for the transformation logic.